File: resources/messaging/app_instance.py
import boto3
import logging
import os

logger = logging.getLogger()
try:
    log_level = os.environ["LogLevel"]
    if log_level not in ["INFO", "DEBUG"]:
        log_level = "INFO"
except:
    log_level = "INFO"
logger.setLevel(log_level)

chime = boto3.client("chime-sdk-identity")
ssm = boto3.client("ssm")


def create_messaging_app_instance(
    uid,
    name=None,
    metadata=None,
    clientRequestToken=None,
    **kwargs,
):
    logger.info(f"Creating a new Amazon Chime SDK Messaging App Instance")

    params = {}
    params["Name"] = name
    if metadata:
        params["Metadata"] = metadata
    if clientRequestToken:
        params["ClientRequestToken"] = clientRequestToken
    logger.info(f"Params to use: {params}")
    try:
        app_instance_arn = chime.create_app_instance(**params)["AppInstanceArn"]
    except Exception as e:
        error = {"error": f"Exception thrown: {e}"}
        logger.error(error)
        raise RuntimeError(error)

    try:
        ssm.put_parameter(
            Name="/chime/appInstanceArn/" + uid,
            Description="appInstanceArn",
            Overwrite=True,
            Value=app_instance_arn,
            Type="String",
        )
    except Exception as e:
        error = {"error": f"Exception thrown: {e}"}
        logger.error(error)
        raise RuntimeError(error)

    return app_instance_arn


def delete_messaging_app_instance(uid):
    logger.info(f"Deleting an Amazon Chime SDK Messaging App Instance")
    try:
        app_instance_to_delete = ssm.get_parameter(Name="/chime/appInstanceArn/" + str(uid),)[
            "Parameter"
        ]["Value"]
    except Exception as e:
        error = {"error": f"Exception thrown: {e}"}
        logger.error(error)
        raise RuntimeError(error)
    try:
        chime.delete_app_instance(AppInstanceArn=app_instance_to_delete)
    except Exception as e:
        error = {"error": f"Exception thrown: {e}"}
        logger.error(error)
        raise RuntimeError(error)

    return True

File: src/resources/messaging/appInstance.ts
/* eslint-disable import/no-extraneous-dependencies */
import {
  ChimeSDKIdentityClient,
  CreateAppInstanceCommand,
  CreateAppInstanceCommandInput,
  CreateAppInstanceCommandOutput,
  DeleteAppInstanceCommand,
  Tag,
} from '@aws-sdk/client-chime-sdk-identity';

import {
  SSMClient,
  DeleteParameterCommand,
  GetParameterCommand,
  GetParameterCommandOutput,
  PutParameterCommand,
} from '@aws-sdk/client-ssm';

const chimeClient = new ChimeSDKIdentityClient({
  region: process.env.AWS_REGION,
});

const ssmClient = new SSMClient({ region: process.env.AWS_REGION });

let createAppInstanceCommandParams: CreateAppInstanceCommandInput;
let createAppInstanceCommandResponse: CreateAppInstanceCommandOutput;
let getParameterCommandOutput: GetParameterCommandOutput;
let updatedTags: Tag[];

interface Tags {
  key: string;
  value: string;
}

interface CreateAppInstanceProps {
  name?: string;
  metadata?: string;
  clientRequestToken?: string;
  tags?: Tags[];
}

export const CreateAppInstance = async (
  uid: string,
  props: CreateAppInstanceProps,
) => {
  updatedTags = [];
  if (props.tags) {
    props.tags.forEach((tag) => {
      updatedTags.push({ Key: tag.key, Value: tag.value });
    });
  }

  createAppInstanceCommandParams = {
    Name: props.name,
    ...(props.metadata && { Metadata: props.metadata }),
    ...(props.clientRequestToken && {
      ClientRequestToken: props.clientRequestToken,
    }),
    ...(updatedTags.length > 0 && { Tags: updatedTags }),
  };

  try {
    createAppInstanceCommandResponse = await chimeClient.send(
      new CreateAppInstanceCommand(createAppInstanceCommandParams),
    );
  } catch (error) {
    if (error instanceof Error) {
      console.error(error);
      throw error;
    }
  }

  try {
    await ssmClient.send(
      new PutParameterCommand({
        Name: `/chime/appInstanceArn/${uid}`,
        Description: 'App Instance Arn',
        Value: createAppInstanceCommandResponse.AppInstanceArn,
        Overwrite: true,
        Type: 'String',
      }),
    );
  } catch (error) {
    if (error instanceof Error) {
      console.error(error);
      throw error;
    }
  }

  return {
    appInstanceArn: createAppInstanceCommandResponse.AppInstanceArn,
  };
};

export const DeleteAppInstance = async (uid: string) => {
  try {
    getParameterCommandOutput = await ssmClient.send(
      new GetParameterCommand({ Name: `/chime/appInstanceArn/${uid}` }),
    );
  } catch (error) {
    if (error instanceof Error) {
      console.error(error);
      throw error;
    }
  }

  try {
    await chimeClient.send(
      new DeleteAppInstanceCommand({
        AppInstanceArn: getParameterCommandOutput.Parameter!.Value,
      }),
    );
    await ssmClient.send(
      new DeleteParameterCommand({ Name: `/chime/appInstanceArn/${uid}` }),
    );
  } catch (error) {
    if (error instanceof Error) {
      console.error(error);
      throw error;
    }
  }
};
